routing/edge_estimator.cpp

```diff
@@ -17,33 +32,60 @@ inline double TimeBetweenSec(m2::PointD const & from, m2::PointD const & to, dou
 class CarEdgeEstimator : public EdgeEstimator
 {
 public:
-  CarEdgeEstimator(IVehicleModel const & vehicleModel);
+  CarEdgeEstimator(IVehicleModel const & vehicleModel, traffic::TrafficInfoGetter const & getter);
 
   // EdgeEstimator overrides:
-  double CalcEdgesWeight(RoadGeometry const & road, uint32_t pointFrom,
+  void Start(MwmSet::MwmId const & mwmId) override;
+  void Finish() override;
+  double CalcEdgesWeight(uint32_t featureId, RoadGeometry const & road, uint32_t pointFrom,
                          uint32_t pointTo) const override;
   double CalcHeuristic(m2::PointD const & from, m2::PointD const & to) const override;
 
 private:
+  TrafficInfoGetter const & m_trafficGetter;
+  shared_ptr<traffic::TrafficInfo> m_trafficInfo;
   double const m_maxSpeedMPS;
 };
 
-CarEdgeEstimator::CarEdgeEstimator(IVehicleModel const & vehicleModel)
-  : m_maxSpeedMPS(vehicleModel.GetMaxSpeed() * kKMPH2MPS)
+CarEdgeEstimator::CarEdgeEstimator(IVehicleModel const & vehicleModel,
+                                   traffic::TrafficInfoGetter const & getter)
+  : m_trafficGetter(getter), m_maxSpeedMPS(vehicleModel.GetMaxSpeed() * kKMPH2MPS)
+{
+}
+
+void CarEdgeEstimator::Start(MwmSet::MwmId const & mwmId)
+{
+  m_trafficInfo = m_trafficGetter.GetTrafficInfo(mwmId);
+}
+
+void CarEdgeEstimator::Finish()
 {
+  m_trafficInfo.reset();
 }
 
-double CarEdgeEstimator::CalcEdgesWeight(RoadGeometry const & road, uint32_t pointFrom,
-                                         uint32_t pointTo) const
+double CarEdgeEstimator::CalcEdgesWeight(uint32_t featureId, RoadGeometry const & road,
+                                         uint32_t pointFrom, uint32_t pointTo) const
 {
   uint32_t const start = min(pointFrom, pointTo);
   uint32_t const finish = max(pointFrom, pointTo);
   ASSERT_LESS(finish, road.GetPointsCount(), ());
 
   double result = 0.0;
   double const speedMPS = road.GetSpeed() * kKMPH2MPS;
+  auto const dir = pointFrom < pointTo ? TrafficInfo::RoadSegmentId::kForwardDirection
+                                       : TrafficInfo::RoadSegmentId::kReverseDirection;
   for (uint32_t i = start; i < finish; ++i)
-    result += TimeBetweenSec(road.GetPoint(i), road.GetPoint(i + 1), speedMPS);
+  {
+    double edgeWeight = TimeBetweenSec(road.GetPoint(i), road.GetPoint(i + 1), speedMPS);
+    if (m_trafficInfo)
+    {
+      SpeedGroup const speedGroup =
+          m_trafficInfo->GetSpeedGroup(TrafficInfo::RoadSegmentId(featureId, i, dir));
+      CHECK_LESS(speedGroup, SpeedGroup::Count, ());
```

Comment:
Feel free to replace CHECK by ASSERT here, as traffic info should always return a valid speed group by contract.